File: Kanban.java
// Kanban.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

enum Status {
    TODO("To Do"),
    IN_PROGRESS("In Progress"),
    DONE("Done");

    private final String display;
    Status(String display) { this.display = display; }
    public String getDisplay() { return display; }

    public static Status fromDisplay(String display) {
        for (Status s : values()) {
            if (s.display.equals(display)) return s;
        }
        throw new IllegalArgumentException("Неизвестный статус");
    }
}

enum Priority {
    LOW("Низкий"),
    MEDIUM("Средний"),
    HIGH("Высокий");

    private final String display;
    Priority(String display) { this.display = display; }
    public String getDisplay() { return display; }

    public static Priority fromDisplay(String display) {
        for (Priority p : values()) {
            if (p.display.equals(display)) return p;
        }
        return MEDIUM;
    }
}

record Task(int id, String title, String description, Status status, Priority priority, String createdAt) implements Serializable {
    public Task {
        Objects.requireNonNull(title);
        Objects.requireNonNull(description);
        Objects.requireNonNull(status);
        Objects.requireNonNull(priority);
    }
}

class KanbanBoard implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<Task> tasks = new ArrayList<>();
    private int nextId = 1;

    public Task addTask(String title, String desc, Priority priority) {
        Task task = new Task(nextId, title, desc, Status.TODO, priority,
                LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        tasks.add(task);
        nextId++;
        return task;
    }

    public Optional<Task> findTask(int id) {
        return tasks.stream().filter(t -> t.id() == id).findFirst();
    }

    public boolean moveTask(int id, Status newStatus) {
        return findTask(id).map(t -> {
            tasks.remove(t);
            Task updated = new Task(t.id(), t.title(), t.description(), newStatus, t.priority(), t.createdAt());
            tasks.add(updated);
            return true;
        }).orElse(false);
    }

    public boolean deleteTask(int id) {
        return tasks.removeIf(t -> t.id() == id);
    }

    public List<Task> filterByStatus(Status status) {
        return tasks.stream().filter(t -> t.status() == status).collect(Collectors.toList());
    }

    public void saveToFile(String filename) throws IOException {
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(this);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            KanbanBoard loaded = (KanbanBoard) ois.readObject();
            this.tasks = loaded.tasks;
            this.nextId = loaded.nextId;
        }
    }

    public List<Task> getTasks() { return Collections.unmodifiableList(tasks); }
}

public class Kanban {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static void printTask(Task task) {
        String emoji = switch (task.status()) {
            case TODO -> "📌";
            case IN_PROGRESS -> "🔄";
            case DONE -> "✅";
        };
        System.out.printf("%s #%d - %s (%s)%n", emoji, task.id(), task.title(), task.status().getDisplay());
        System.out.printf("   Описание: %s%n", task.description());
        System.out.printf("   Приоритет: %s, Создана: %s%n", task.priority().getDisplay(), task.createdAt());
    }

    public static void main(String[] args) {
        KanbanBoard board = new KanbanBoard();
        try {
            board.loadFromFile("kanban_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные, начинаем с пустой доски.");
        }

        while (true) {
            System.out.println("\n===== КАНБАН-ДОСКА (Java) =====");
            System.out.println("1. Показать все задачи");
            System.out.println("2. Добавить задачу");
            System.out.println("3. Переместить задачу (изменить статус)");
            System.out.println("4. Удалить задачу");
            System.out.println("5. Сохранить в файл");
            System.out.println("6. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    if (board.getTasks().isEmpty()) {
                        System.out.println("Задач нет.");
                    } else {
                        board.getTasks().forEach(Kanban::printTask);
                    }
                }
                case "2" -> {
                    String title = readString("Введите название: ");
                    if (title.isBlank()) {
                        System.out.println("Название не может быть пустым.");
                        continue;
                    }
                    String desc = readString("Введите описание: ");
                    System.out.println("Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий");
                    String prio = readString("> ");
                    Priority priority = switch (prio) {
                        case "1" -> Priority.LOW;
                        case "3" -> Priority.HIGH;
                        default -> Priority.MEDIUM;
                    };
                    Task task = board.addTask(title, desc, priority);
                    System.out.printf("Задача добавлена с ID %d%n", task.id());
                }
                case "3" -> {
                    int id = readInt("Введите ID задачи: ");
                    var opt = board.findTask(id);
                    if (opt.isEmpty()) {
                        System.out.println("Задача не найдена.");
                        continue;
                    }
                    Task task = opt.get();
                    System.out.printf("Текущий статус: %s%n", task.status().getDisplay());
                    System.out.println("Выберите новый статус: 1-To Do, 2-In Progress, 3-Done");
                    String newChoice = readString("> ");
                    Status newStatus = switch (newChoice) {
                        case "1" -> Status.TODO;
                        case "2" -> Status.IN_PROGRESS;
                        case "3" -> Status.DONE;
                        default -> null;
                    };
                    if (newStatus == null) {
                        System.out.println("Неверный выбор.");
                        continue;
                    }
                    if (board.moveTask(id, newStatus)) {
                        System.out.println("Статус обновлён.");
                    } else {
                        System.out.println("Ошибка.");
                    }
                }
                case "4" -> {
                    int id = readInt("Введите ID задачи для удаления: ");
                    if (board.deleteTask(id)) {
                        System.out.println("Задача удалена.");
                    } else {
                        System.out.println("Задача не найдена.");
                    }
                }
                case "5" -> {
                    try {
                        board.saveToFile("kanban_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "6" -> {
                    try {
                        board.loadFromFile("kanban_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}

File: kanban.go
// kanban.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Status string

const (
	TODO       Status = "To Do"
	IN_PROGRESS Status = "In Progress"
	DONE       Status = "Done"
)

type Priority string

const (
	LOW    Priority = "Низкий"
	MEDIUM Priority = "Средний"
	HIGH   Priority = "Высокий"
)

type Task struct {
	ID          int       `json:"id"`
	Title       string    `json:"title"`
	Description string    `json:"description"`
	Status      Status    `json:"status"`
	Priority    Priority  `json:"priority"`
	CreatedAt   time.Time `json:"created_at"`
}

type KanbanBoard struct {
	tasks   []Task
	nextID  int
}

func NewKanbanBoard() *KanbanBoard {
	return &KanbanBoard{
		tasks:  []Task{},
		nextID: 1,
	}
}

func (b *KanbanBoard) AddTask(title, desc string, priority Priority) Task {
	task := Task{
		ID:          b.nextID,
		Title:       title,
		Description: desc,
		Status:      TODO,
		Priority:    priority,
		CreatedAt:   time.Now(),
	}
	b.tasks = append(b.tasks, task)
	b.nextID++
	return task
}

func (b *KanbanBoard) FindTask(id int) *Task {
	for i := range b.tasks {
		if b.tasks[i].ID == id {
			return &b.tasks[i]
		}
	}
	return nil
}

func (b *KanbanBoard) MoveTask(id int, newStatus Status) bool {
	task := b.FindTask(id)
	if task == nil {
		return false
	}
	task.Status = newStatus
	return true
}

func (b *KanbanBoard) DeleteTask(id int) bool {
	for i, t := range b.tasks {
		if t.ID == id {
			b.tasks = append(b.tasks[:i], b.tasks[i+1:]...)
			return true
		}
	}
	return false
}

func (b *KanbanBoard) SaveToFile(filename string) error {
	data, err := json.MarshalIndent(b.tasks, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, data, 0644)
}

func (b *KanbanBoard) LoadFromFile(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var loaded []Task
	if err := json.Unmarshal(data, &loaded); err != nil {
		return err
	}
	b.tasks = loaded
	for _, t := range loaded {
		if t.ID >= b.nextID {
			b.nextID = t.ID + 1
		}
	}
	return nil
}

func printTask(task Task) {
	emoji := map[Status]string{TODO: "📌", IN_PROGRESS: "🔄", DONE: "✅"}
	fmt.Printf("%s #%d - %s (%s)\n", emoji[task.Status], task.ID, task.Title, task.Status)
	fmt.Printf("   Описание: %s\n", task.Description)
	fmt.Printf("   Приоритет: %s, Создана: %s\n", task.Priority, task.CreatedAt.Format("2006-01-02 15:04:05"))
}

func readString(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func readInt(prompt string) int {
	for {
		input := readString(prompt)
		if val, err := strconv.Atoi(input); err == nil {
			return val
		}
		fmt.Println("Введите число.")
	}
}

func main() {
	board := NewKanbanBoard()
	if err := board.LoadFromFile("kanban_data.json"); err != nil {
		fmt.Println("Ошибка загрузки:", err)
	}

	for {
		fmt.Println("\n===== КАНБАН-ДОСКА (Go) =====")
		fmt.Println("1. Показать все задачи")
		fmt.Println("2. Добавить задачу")
		fmt.Println("3. Переместить задачу (изменить статус)")
		fmt.Println("4. Удалить задачу")
		fmt.Println("5. Сохранить в файл")
		fmt.Println("6. Загрузить из файла")
		fmt.Println("0. Выход")
		choice := readString("Выберите действие: ")

		switch choice {
		case "0":
			return
		case "1":
			if len(board.tasks) == 0 {
				fmt.Println("Задач нет.")
			} else {
				for _, t := range board.tasks {
					printTask(t)
				}
			}
		case "2":
			title := readString("Введите название: ")
			if title == "" {
				fmt.Println("Название не может быть пустым.")
				continue
			}
			desc := readString("Введите описание: ")
			fmt.Println("Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий")
			prio := readString("> ")
			var priority Priority
			switch prio {
			case "1":
				priority = LOW
			case "3":
				priority = HIGH
			default:
				priority = MEDIUM
			}
			task := board.AddTask(title, desc, priority)
			fmt.Printf("Задача добавлена с ID %d\n", task.ID)
		case "3":
			id := readInt("Введите ID задачи: ")
			task := board.FindTask(id)
			if task == nil {
				fmt.Println("Задача не найдена.")
				continue
			}
			fmt.Printf("Текущий статус: %s\n", task.Status)
			fmt.Println("Выберите новый статус: 1-To Do, 2-In Progress, 3-Done")
			newChoice := readString("> ")
			var newStatus Status
			switch newChoice {
			case "1":
				newStatus = TODO
			case "2":
				newStatus = IN_PROGRESS
			case "3":
				newStatus = DONE
			default:
				fmt.Println("Неверный выбор.")
				continue
			}
			if board.MoveTask(id, newStatus) {
				fmt.Println("Статус обновлён.")
			} else {
				fmt.Println("Ошибка.")
			}
		case "4":
			id := readInt("Введите ID задачи для удаления: ")
			if board.DeleteTask(id) {
				fmt.Println("Задача удалена.")
			} else {
				fmt.Println("Задача не найдена.")
			}
		case "5":
			if err := board.SaveToFile("kanban_data.json"); err != nil {
				fmt.Println("Ошибка сохранения:", err)
			} else {
				fmt.Println("Сохранено.")
			}
		case "6":
			if err := board.LoadFromFile("kanban_data.json"); err != nil {
				fmt.Println("Ошибка загрузки:", err)
			} else {
				fmt.Println("Загружено.")
			}
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}
